File: backend/app/models/dcf_valuation.py
from datetime import datetime, timezone
from sqlalchemy import Column, Integer, String, Float, Text, DateTime
from app.database import Base


class DCFValuation(Base):
    __tablename__ = "dcf_valuations"

    id = Column(Integer, primary_key=True, index=True)

    # Identification
    ticker = Column(String(20), nullable=True, index=True)
    company_name = Column(String(255), nullable=True)

    # Core inputs
    base_fcf = Column(Float, nullable=False)
    projection_years = Column(Integer, default=5)
    growth_rate_phase1 = Column(Float, nullable=False)
    growth_rate_phase2 = Column(Float, nullable=False)
    discount_rate = Column(Float, nullable=False)
    terminal_method = Column(String(20), default="gordon")
    terminal_multiple = Column(Float, nullable=True)

    # WACC components
    risk_free_rate = Column(Float, nullable=True)
    equity_risk_premium = Column(Float, nullable=True)
    beta = Column(Float, nullable=True)
    cost_of_equity = Column(Float, nullable=True)
    cost_of_debt = Column(Float, nullable=True)
    tax_rate = Column(Float, nullable=True)
    debt_weight = Column(Float, nullable=True)
    equity_weight = Column(Float, nullable=True)

    # Balance sheet adjustments
    total_debt = Column(Float, nullable=True)
    total_cash = Column(Float, nullable=True)
    shares_outstanding = Column(Float, nullable=True)
    current_price = Column(Float, nullable=True)

    # Computed results
    enterprise_value = Column(Float, nullable=True)
    equity_value = Column(Float, nullable=True)
    intrinsic_value_per_share = Column(Float, nullable=True)
    upside_downside_pct = Column(Float, nullable=True)
    margin_of_safety = Column(Float, nullable=True)
    terminal_value = Column(Float, nullable=True)

    # JSON blobs
    projected_fcfs = Column(Text, nullable=True)
    sensitivity_table = Column(Text, nullable=True)
    valuation_verdict = Column(String(30), nullable=True)

    created_at = Column(
        DateTime, default=lambda: datetime.now(timezone.utc), index=True
    )

File: laravel-backend/app/Models/DCFValuation.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class DCFValuation extends Model
{
    public $timestamps = false;

    protected $table = 'dcf_valuations';

    protected $fillable = [
        'ticker',
        'company_name',
        'base_fcf',
        'projection_years',
        'growth_rate_phase1',
        'growth_rate_phase2',
        'discount_rate',
        'terminal_method',
        'terminal_multiple',
        'risk_free_rate',
        'equity_risk_premium',
        'beta',
        'cost_of_equity',
        'cost_of_debt',
        'tax_rate',
        'debt_weight',
        'equity_weight',
        'total_debt',
        'total_cash',
        'shares_outstanding',
        'current_price',
        'enterprise_value',
        'equity_value',
        'intrinsic_value_per_share',
        'upside_downside_pct',
        'margin_of_safety',
        'terminal_value',
        'projected_fcfs',
        'sensitivity_table',
        'valuation_verdict',
    ];

    protected $casts = [
        'base_fcf' => 'double',
        'growth_rate_phase1' => 'double',
        'growth_rate_phase2' => 'double',
        'discount_rate' => 'double',
        'terminal_multiple' => 'double',
        'risk_free_rate' => 'double',
        'equity_risk_premium' => 'double',
        'beta' => 'double',
        'cost_of_equity' => 'double',
        'cost_of_debt' => 'double',
        'tax_rate' => 'double',
        'debt_weight' => 'double',
        'equity_weight' => 'double',
        'total_debt' => 'double',
        'total_cash' => 'double',
        'shares_outstanding' => 'double',
        'current_price' => 'double',
        'enterprise_value' => 'double',
        'equity_value' => 'double',
        'intrinsic_value_per_share' => 'double',
        'upside_downside_pct' => 'double',
        'margin_of_safety' => 'double',
        'terminal_value' => 'double',
        'created_at' => 'datetime',
    ];

    public function getProjectedFcfsArrayAttribute(): array
    {
        return json_decode($this->projected_fcfs, true) ?? [];
    }

    public function getSensitivityTableArrayAttribute(): array
    {
        return json_decode($this->sensitivity_table, true) ?? [];
    }
}
